### DISCUSSION TITLE
A cordial invite

### DISCUSSION THREAD

**Discussion by Ivorforce** [j20001970/GDMP#55] [General]
Hi there!
I found your repo when somebody on the Godot Discord server was working with it. Exciting stuff!
I'm part of an initiative with the goal of bringing scientific computation to Godot users. To be specific, so far we've brought together a [NumPy-like tensor computation library called NumDot](https://github.com/Ivorforce/NumDot), and [OpenCV-](https://github.com/matt-s-clark/godot-gdextension-opencv) as well as [IREE bindings](https://github.com/iree-gd/iree.gd) for GDScript.

As an author of a very relevant bindings library, we'd be happy to see you join and come chat as well! We're interested in bringing like-minded people together, create interoperable libraries, and generally join efforts. If you're interested, feel free to come directly to our server: https://discord.gg/mwS2sW6V5M

**Comment by j20001970** [comment]
Thank you for the passionate invitation, it's nice to see many projects building around Godot ecosystem that are not only specific to games, but also for broader applications usage.

Unfortunately, I do not feel like joining Discord servers for the purpose of project development at the moment. That said, I'm open to the ideas of interoperability, as it is quite an important aspect for an open, strong software environment. Please feel free to discuss on the repo where GDMP project could improve in this regard if you see it fit.

Thanks again for the thoughtful invitation, hope community efforts around the world continue to grow and prosper.

**Reply by Ivorforce** [reply]
> I do not feel like joining Discord servers for the purpose of project development at the moment

Alright, that's fair. But feel free to ping me if you've got any questions or ideas :)

We're currently getting the OpenCV bindings out the door, and will then implement on a general purpose data interchange protocol, analogous to the [Python buffer protocol](https://docs.python.org/3/c-api/buffer.html). If you got any comments for that, let me know. Otherwise, we'll let you know when we got something.

**Reply by GeorgeS2019** [reply]
@Ivorforce 
I hope you are a trained academics to understand that before embarking on a direction, you have to do the necessary due diligent to survey if the decision you have recommended e.g. IREE or NumDot, if that help to secure a new grant due to something have not attempted but too early stage to lobby Godot communities.

Have you asked yourself, what exactly the state of Godot communities?

What exactly the Godot communities need right now?

Are your proposals align with helping the Godot communities moving forwards.?

**Reply by Ivorforce** [reply]
@GeorgeS2019 All projects and people currently in the initiative are doing this in their free time. There are no grants, surveys or other actual research involved. In other words, it is a passion project completely disjunct from the academic careers some of us are busy with at their day jobs :)

**Reply by GeorgeS2019** [reply]
I will check the GodotCon 2024 presentations first and then I can offer you a better feedback.

Again, for scientific computation, godot4 support python and effort needed to do a PoC which of the Godot4 python GDExtension provides the best environment for python driven Scientific computation. 

This is the boring part, usually for developers. However, this is exactly the core part that help to decide how to steer scientific computation for godot. This remark will anger many developers who just want to try something new, without doing the due diligent of addressing who are your customers and how best your software engineering talents can be put to deliver the best impact for the society.

**Reply by Ivorforce** [reply]
I appreciate starting discourse about alternate or future directions for a better scientific workflow in the Godot ecosystem. This is discussion we should be having! 
I don't think this Forum is necessarily the best place for it, though. If you want to have a chat, the above linked Discord would be better suited. Or feel free to suggest another space.

**Reply by Ivorforce** [reply]
Hey again @GeorgeS2019, we've discussed your comments and they made us think. We ended up putting together a survey to gain better perspective on the Godot community, and the need for compute related libraries. Find it here: https://docs.google.com/forms/d/e/1FAIpQLScLlxEiZ5DevStUnRGHeGyN98AkwlXmzzUlguKMPWfNu-vodw/viewform

We have not published the form anywhere yet, and would be very open to feedback before sending it around. If you've got any comments, let us know. The [Discord invitation](https://discord.gg/mwS2sW6V5M) is of course also still open!

**Reply by GeorgeS2019** [reply]
@Ivorforce 
There is no doubt, you and your like minded persons are in the right direction. 
What you all are looking for are using **blender** for scientific workflow. Blender is chosen obviously it supports python and its vast number of ready to use python libraries. 

Instead of blender, you use Godot. 

I hope this helps. Keep it up.

**Comment by j20001970** [comment]
> We're currently getting the OpenCV bindings out the door, and will then implement on a general purpose data interchange protocol, analogous to the [Python buffer protocol](https://docs.python.org/3/c-api/buffer.html).

Curious that what kind of data would like to be interchangeable on OpenCV side, will using Godot built-in type to represent the data be feasible?

**Reply by Ivorforce** [reply]
Exchanging data via Godot types (e.g. PackedByteArray) is a low- tech solution we are considering.
But we also want to support a high performance solution by exchanging buffer pointers (and shapes / strides) directly.

NumDot works with N dimensional tensors and will want to exchange any kind of data of any data type.

OpenCV works primarily on images and will thus primarily be exchanging 2D image data in various data channel configurations.

**Comment by Ivorforce** [comment]
Hey again @j20001970, we have linked your project on a new Awesome Godot Scientific page we have started curating: https://github.com/Ivorforce/Awesome-Godot-Scientific/

Let me know if we should remove your entry, or reword it!
